### DISCUSSION TITLE
Modeling the viscoelastic-wave equation in frequency domain

### DISCUSSION THREAD

**Discussion by aaelmeli** [idaholab/moose#20266] [Q&A Modules: Solid mechanics] [answered]
Hi
Let's assume we have a complex vector field (displacements)` u = u_r+i*u_i` where `u_r` is a vector containing the real parts of `u` and `u_i` is a vector containing imaginary parts of `u`, and `i=sqrt(-1)` . 

The elastic wave equation in the frequency domain can be described as:
```
ρω^2 u_r=∇.σ_r (1)
ρω^2 u_i=∇.σ_i  (2)
```
In that case, the real and imaginary parts of `u` were uncoupled everywhere except on the boundaries (only if we have absorbing boundary conditions). I could model those equations within moose, where the kernel is as follows:
```
[Kernels]
    [urealx]
        type = StressDivergenceTensors
        variable = uxr
        displacements='uxr uyr'
        component = 0
        base_name = real
        volumetric_locking_correction=True
    []
    [urealy]
        type = StressDivergenceTensors
        variable = uyr
        displacements='uxr uyr'
        component = 1
        base_name = real
        volumetric_locking_correction=True
    []
    [uimagx]
        type = StressDivergenceTensors
        variable = uxi
        displacements='uxi uyi'
        component = 0
        base_name = imaginary
        volumetric_locking_correction=True
    []
    [uimagy]
        type = StressDivergenceTensors
        variable = uyi
        displacements='uxi uyi'
        component = 1
        base_name = imaginary
        volumetric_locking_correction=True
    []
    [reaction_realx]
        type = Reaction
        variable = uxr
        rate = ${fparse -omega*omega} 
    []
    [reaction_realy]
        type = Reaction
        variable = uyr
        rate = ${fparse -omega*omega} 
    []
    [reaction_imagx]
        type = Reaction
        variable = uxi
        rate = ${fparse -omega*omega} 
    []
    [reaction_imagy]
        type = Reaction
        variable = uyi
        rate = ${fparse -omega*omega} 
    []
[]
```
Now, I need to add the viscoelasticity to the model, where the elasticity tensor is no longer real-valued, eg. `E = E_r + iωE_i`. Thus, equations `(1)` and `(2)` turn to be:
```
ρω^2 u_r=∇.σ_rr - ∇.σ_ii   (3)
ρω^2 u_i=∇.σ_ir + ∇.σ_ri  (4)
```
with `σ_rr` is the stress associated with the real part of the elasticity tensor `E_r` and the real part of the displacement `u_r`, and `σ_ir` is the stresses associated with the imaginary part of the elasticity tensor `E_i` and the real part of the displacement `u_r`, and vice-versa.

So, the question now is, Can I model equations (3) and (4) in moose? 

Any thoughts will be appreciated.

**Comment by GiudGiud** [comment]
Hello

You need to have a kernel that does the same work as a stressDivergenceTensor (except for the sign), but with a variable that is not the same as the equation name. 

What prevents you from using that one directly is the variable argument is both:
- what gets put through the divergence to compute the contribution to the residual/jacobian
- the name of the equation that this term gets added to

So unless this already exists (@jiangwen84 do you know?), you ll just need to add an argument to this kernel to allow an arbitrary variable `v` to be used in the computation (and if `v` is not provided, then default to using `variable`). 

EDIT: I m actually not sure this isnt already supported. Would just specifying:
```
        variable = uxr
        displacements='uxi uyi'
```
not work?

Guillaume

**Reply by GiudGiud** [reply]
also are there two displaced meshes? Or are all the equations written for the undisplaced mesh?

**Reply by aaelmeli** [reply]
> also are there two displaced meshes? Or are all the equations written for the undisplaced mesh?

The equations are for undisplaced mesh. 


> EDIT: I m actually not sure this isnt already supported. Would just specifying:

Thanks for the suggestion, I will try this and get back to you.

**Reply by aaelmeli** [reply]
> EDIT: I m actually not sure this isnt already supported. Would just specifying:
> 
> ```
>         variable = uxr
>         displacements='uxi uyi'
> ```
> 
> not work?



Hi,

I have tried that option, please see the definition of the kernel and the materials. So, it is a 2D mesh that represents a cantilever, fixed on the left and imposed unit Dirichlet B.C on the right. 
```
[Kernels]
    [urealx_r] # the (real_r) means that stress inteded to be computed is sigma_rr
    type = StressDivergenceTensors
    variable = uxr
    displacements='uxr uyr'
    component = 0
    base_name = real
    []
    [urealx_i] #the (real_i) means that stress inteded to be computed is sigma_ri
    type = StressDivergenceTensors
    variable = uxr
    displacements='uxi uyi'
    component = 0
    base_name = real #do I need to change the material??
    []    
    [urealy_r]
    type = StressDivergenceTensors
    variable = uyr
    displacements='uxr uyr'
    component = 1
    base_name = real
    []
    [urealy_i]
    type = StressDivergenceTensors
    variable = uyr
    displacements='uxi uyi'
    component = 1
    base_name = real
    []    
    [uimagx_i]
    type = StressDivergenceTensors
    variable = uxi
    displacements='uxi uyi'
    component = 0
    base_name = imaginary
    []
    [uimagx_r]
    type = StressDivergenceTensors
    variable = uxi
    displacements='uxr uyr'
    component = 0
    base_name = imaginary
    []    
    [uimagy_i]
    type = StressDivergenceTensors
    variable = uyi
    displacements='uxi uyi'
    component = 1
    base_name = imaginary
    []
    [uimagy_r]
    type = StressDivergenceTensors
    variable = uyi
    displacements='uxr uyr'
    component = 1
    base_name = imaginary
    []    
    [reaction_realx]
        type = Reaction
        variable = uxr
        rate = -10000
    []
    [reaction_realy]
        type = Reaction
        variable = uyr
        rate = -10000
    []
    [reaction_imagx]
        type = Reaction
        variable = uxi
        rate = -10000
    []
    [reaction_imagy]
        type = Reaction
        variable = uyi
        rate = -10000
    []

[]
```
```
Materials]

  [elasticity_tensor_A]
    type = ComputeIsotropicElasticityTensor
    base_name = real
    youngs_modulus = 1
    poissons_ratio = 0.0
  []
  [strain_A]
    type = ComputePlaneSmallStrain
    base_name = real
    displacements='uxr uyr'
  []
  [stress_A]
    type = ComputeLinearElasticStress
    base_name = real
  []

   [elasticity_tensor_B]
    type = ComputeIsotropicElasticityTensor
    base_name = imaginary
    youngs_modulus = 0.1
    poissons_ratio = 0.0
  []
  [strain_B]
    type = ComputePlaneSmallStrain
    base_name = imaginary
    displacements='uxi uyi'
  []
  [stress_B]
    type = ComputeLinearElasticStress
    base_name = imaginary
  []

[]
```
Given this input, I am not getting the expected solution, which I am not sure about the reason for getting this, is it because of the `-ve sign in equation (3)`, or incorrect modeling, or **both**.
One thing that I am wondering, Do I have modify the base name in the kernel definition in one way or another? And, is there a way that I can try to include the `-ve sign` in the kernel definition?, just to try your idea.

**Reply by GiudGiud** [reply]
You could add a boolean parameter to this kernel just to allow the switching of signs. 

I dont think you need to modify the base name.

I have not checked the kernel definition in detail btw, this is really just a try.

**Reply by aaelmeli** [reply]
> You could add a boolean parameter to this kernel just to allow the switching of signs.
> 
> I dont think you need to modify the base name.
> 
> I have not checked the kernel definition in detail btw, this is really just a try.

Thanks for the suggestion, I will do that. 
One question, is it possible to define the youngs_modulus with the `-ve sign`? I wanted to push the sign in the elasticity tensor. I have tried  it but it returned an error (which makes sense).  I know it is not the optimum way of doing it, but I was just thinking of the easiest way to do it.

**Reply by aaelmeli** [reply]
Hi, @GiudGiud 

I have done several trials to use the` StressDivergenceTensors` kernels to model the former equations, `(3) and (4) `with no success. As you have suggested, I have added a `boolean parameter` to switch the sign of the stress divergence kernel. Also, based on what I have understood from the code, "`displacements`" is an array of coupled variables. Thus, I have considered this when writing my inputs. The input `Kernel`, `Material ` and `Executioner `definitions that I believe "may" make sense are as follow:

```
[Kernels]
    [urealx_r] #sigma_rr
    type = StressDivergenceTensors
    variable = uxr
    displacements='uxr uyr'
    component = 0
    base_name = real
    []
    [urealy_r]
    type = StressDivergenceTensors
    variable = uyr
    displacements='uxr uyr'
    component = 1
    base_name = real
    []  
#2
    [uimagx_i] # (-ve)sigma_ii
    type = StressDivergenceTensors
    variable = uxr
    displacements='uxi uyi'
    component = 0
    base_name = imaginary
    negative_stress_divergence='true' #a parameter to negate the stressdivergence term
    []
    [uimagy_i]
    type = StressDivergenceTensors
    variable = uyr
    displacements='uxi uyi'
    component = 1
    base_name = imaginary
    negative_stress_divergence='true'
    [] 
#3
    [urealx_i] #sigma_ri
    type = StressDivergenceTensors
    variable = uxi
    displacements='uxr uyr'
    component = 0
    base_name = real_i 
    []   
    [urealy_i]
    type = StressDivergenceTensors
    variable = uyi
    displacements='uxr uyr'
    component = 1
    base_name = real_i
    []  
#4
    [uimagx_r]#sigma_ir
    type = StressDivergenceTensors
    variable = uxi
    displacements='uxi uyi'
    component = 0
    base_name = imaginary_r
    []        
    [uimagy_r]
    type = StressDivergenceTensors
    variable = uyi
    displacements='uxi uyi'
    component = 1
    base_name = imaginary_r
    []
 #reaction terms              
    [reaction_realx]
        type = Reaction
        variable = uxr
        rate = -10000
    []
    [reaction_realy]
        type = Reaction
        variable = uyr
        rate = -10000
    []
    [reaction_imagx]
        type = Reaction
        variable = uxi
        rate = -10000
    []
    [reaction_imagy]
        type = Reaction
        variable = uyi
        rate = -10000
    []
[]

[Materials]

  [elasticity_tensor_real]
    type = ComputeIsotropicElasticityTensor
    base_name = real
    youngs_modulus = 1
    poissons_ratio = 0.0
  []
  [strain_real]
    type = ComputePlaneSmallStrain
    base_name = real
    displacements='uxr uyr'
  []
  [stress_real]
    type = ComputeLinearElasticStress
    base_name = real
  []
  [elasticity_tensor_real_i]
    type = ComputeIsotropicElasticityTensor
    base_name = real_i
    youngs_modulus = 0.1
    poissons_ratio = 0.0
  []
  [strain_real_i]
    type = ComputePlaneSmallStrain
    base_name = real_i
    displacements='uxr uyr'
  []
  [stress_real_i]
    type = ComputeLinearElasticStress
    base_name = real_i
  []
    [elasticity_tensor_imaginary]
    type = ComputeIsotropicElasticityTensor
    base_name = imaginary
    youngs_modulus = 0.1
    poissons_ratio = 0.0
  []
  [strain_imaginary]
    type = ComputePlaneSmallStrain
    base_name = imaginary
    displacements='uxi uyi'
  []
  [stress_imaginary]
    type = ComputeLinearElasticStress
    base_name = imaginary
  []
   [elasticity_tensor_imaginary_r]
    type = ComputeIsotropicElasticityTensor
    base_name = imaginary_r
    youngs_modulus = 1
    poissons_ratio = 0.0
  []
  [strain_imaginary_r]
    type = ComputePlaneSmallStrain
    base_name = imaginary_r
    displacements='uxi uyi'
  []
  [stress_imaginary_r]
    type = ComputeLinearElasticStress
    base_name = imaginary_r
  []

[]

[Executioner]
  type = Steady
  solve_type=LINEAR   
 petsc_options_iname = ' -pc_type'
 petsc_options_value = 'lu'
[]
```
The real and imaginary parts of x-displacements using that input are as follow:
![image](https://user-images.githubusercontent.com/88537645/154780322-f02ea968-c0da-4da2-94cf-30615dd654fa.png)

Where the expected solution should be something like this:
![image](https://user-images.githubusercontent.com/88537645/154780117-e4ef689d-268e-4670-8e5d-36e18af87133.png)

So, I have a couple of questions regarding this,
1- Am I understanding the code correctly with respect to the "displacements" array? are these the test functions? if yes, is the input above correct?
2- Do you think this way of coupling may not be achieved correctly through the `StressDiverenceTensors `and some other kernels e.g. `Diffusion`, may be better?

**Reply by GiudGiud** [reply]
The displacements parameter is usually just that, a measure of how the mesh deformed.

So changing the base_name already made sure you used the right stress tensor. 
I'm actually wondering if changing the displacements was necessary/a good idea. I dont quite understand what this kernel is doing with regards to displacements, so if you do you should look at ``StressDivergenceTensors::computeQpOffDiagJacobian(unsigned int jvar)
`` to see how the displacements should be set.

**Comment by hugary1995** [comment, marked as answer]
I tried to write an input file, and the solution looks correct:
![image](https://user-images.githubusercontent.com/31555204/155055752-e723929a-d947-4a37-bbbf-a3c3465ec56d.png)

I am pasting my input file below, please compare it with yours to see what you were missing. I think you were very close. I modified the `StressDivergenceTensors` to take a coefficient so that I could control its sign, which is similar to your boolean parameter.

```python
[Mesh]
  type = GeneratedMesh
  dim = 1
  xmin = 0
  xmax = 1
  nx = 1000
[]

[Variables]
  [ui]
  []
  [ur]
  []
[]

[BCs]
  [left_real]
    type = DirichletBC
    variable = ur
    value = 0
    boundary = left
  []
  [left_imaginery]
    type = DirichletBC
    variable = ui
    value = 0
    boundary = left
  []
  [right_real]
    type = DirichletBC
    variable = ur
    value = 1
    boundary = right
  []
  [right_imaginery]
    type = DirichletBC
    variable = ui
    value = 0
    boundary = right
  []
[]

[Kernels]
  [real_rr]
    type = StressDivergenceTensors
    variable = ur
    displacements = 'ur'
    component = 0
    base_name = rr
    coefficient = 1
  []
  [real_ii]
    type = StressDivergenceTensors
    variable = ur
    displacements = 'ui'
    component = 0
    base_name = ii
    coefficient = -1
  []
  [real_reaction]
    type = MatReaction
    variable = ur
    mob_name = w
  []
  [imaginery_ir]
    type = StressDivergenceTensors
    variable = ui
    displacements = 'ur'
    component = 0
    base_name = ir
    coefficient = 1
  []
  [imaginery_ri]
    type = StressDivergenceTensors
    variable = ui
    displacements = 'ui'
    component = 0
    base_name = ri
    coefficient = 1
  []
  [imaginery_reaction]
    type = MatReaction
    variable = ui
    mob_name = w
  []
[]

[Materials]
  [bulk]
    type = GenericFunctionMaterial
    prop_names = 'w'
    prop_values = '10000'
  []
  [E_rr]
    type = ComputeIsotropicElasticityTensor
    youngs_modulus = 1
    poissons_ratio = 0.0
    base_name = rr
  []
  [E_ri]
    type = ComputeIsotropicElasticityTensor
    youngs_modulus = 1
    poissons_ratio = 0.0
    base_name = ri
  []
  [E_ir]
    type = ComputeIsotropicElasticityTensor
    youngs_modulus = 0.1
    poissons_ratio = 0.0
    base_name = ir
  []
  [E_ii]
    type = ComputeIsotropicElasticityTensor
    youngs_modulus = 0.1
    poissons_ratio = 0.0
    base_name = ii
  []
  [strain_rr]
    type = ComputeSmallStrain
    displacements = 'ur'
    base_name = rr
  []
  [strain_ri]
    type = ComputeSmallStrain
    displacements = 'ui'
    base_name = ri
  []
  [strain_ir]
    type = ComputeSmallStrain
    displacements = 'ur'
    base_name = ir
  []
  [strain_ii]
    type = ComputeSmallStrain
    displacements = 'ui'
    base_name = ii
  []
  [stress_rr]
    type = ComputeLinearElasticStress
    base_name = rr
  []
  [stress_ri]
    type = ComputeLinearElasticStress
    base_name = ri
  []
  [stress_ir]
    type = ComputeLinearElasticStress
    base_name = ir
  []
  [stress_ii]
    type = ComputeLinearElasticStress
    base_name = ii
  []
[]

[Executioner]
  type = Steady
  solve_type = FD
  petsc_options_iname = '-pc_type'
  petsc_options_value = 'lu'
[]

[Outputs]
  exodus = true
[]
```

**Reply by hugary1995** [reply]
This is apparently not optimal: I had to use four base names to define four different sets of parameters and constitutive models. This is sort of hacking the tensor mechanics to solve your specific problem, and the jacobian of the `StressDivergenceTensors` is also "wrong" with this hack. As a result, it'll have a hard time converging once the viscoelasticity becomes nonlinear. Using automatic differentiation might help you, e.g. `ADStressDivergenceTensors` ad AD materials. Going forward, I suggest you to create your own app.

**Reply by aaelmeli** [reply]
Thank you, @hugary1995, the issue was in the "jacobian", thus, specifying solver type as LINEAR resulted in wrong results. AD helps a lot. Thanks again.
